### PR TITLE
Optimize spill performance by buffering spill data write

### PR DESCRIPTION
**Pull request by xiaoxmeng**
Optimize disk spilling performance for hash build and hash join by
buffering the serialized spill data before persist to disk. With 1MB
buffering per operator, for a given Meta internal production shadowing
query, the spilling performance has been improved a lot. The hash
build cpu time drops from ~10 hrs to <1hr (>10x). The hash probe
cpu time drops from ~1.5hr to ~10 mins (~10x). The spilling stats show
the performance improvement is due to the reduced number of remote
ios. spillDiskWrites of hash build drops from 4M to 100K and spillWriteTime
drops from >7hrs to <3mins.
This PR adds query config to allow user to set per-query write buffer size.
The follow will add the corresponding native session property at Presto to
allow user to configure in Prestissimo.

### CODE REVIEW

File: velox/core/QueryConfig.h [hunk 1]
```diff
@@ -433,6 +440,11 @@ class QueryConfig {
     return get<std::string>(kSpillCompressionKind, "none");
   }
 
+  uint64_t spillWriteBufferSize() const {
+    // The default write buffer size set to 4MB.
```
Comment (mbasmanova) [inline velox/core/QueryConfig.h:444]: Why is the default set to 4MB? The PR description mentions an experiment that used 1MB.

File: velox/docs/configs.rst [hunk 2]
```diff
@@ -198,6 +198,11 @@ Spilling
      - integer
      - 0
      - The maximum allowed spill file size. Zero means unlimited.
+   * - spill_write_buffer_size
+     -integer
```
Comment (mbasmanova) [inline velox/docs/configs.rst:202]: space is missing in "-integer"

File: velox/docs/configs.rst [hunk 3]
```diff
@@ -198,6 +198,11 @@ Spilling
...
+     - 4MB
+     - The maximum size in bytes to buffer the serialized spill data before write to disk for spilling io efficiency.
```
Comment (mbasmanova) [inline velox/docs/configs.rst:204]: "for spilling io efficiency" -> "for IO efficiency"

File: velox/docs/configs.rst [hunk 4]
```diff
@@ -198,6 +198,11 @@ Spilling
...
+       If it is set to zero, then spill write buffering is disabled.
```
Comment (mbasmanova) [inline velox/docs/configs.rst:205]: If set to zero, buffering is disabled.

File: velox/docs/configs.rst [hunk 5]
```diff
@@ -206,6 +211,12 @@ Spilling
        If the limit is zero, then the spiller always spills a previously spilled partition if it has any data. This is
        to avoid spill from a partition with a small amount of data which might result in generating too many small
        spilled files.
+   * - spill_compression_codec
+     - string
+     - none
+     - Specifies the compression algorithm type to compress the spilled data before write to disk to trade cpu for
```
Comment (mbasmanova) [inline velox/docs/configs.rst:217]: This is unrelated change, right? Next time, please, consider making it in a separate PR.

"to trade cpu for spilling io efficiency" -> "to trade CPU for IO efficiency"

File: velox/docs/configs.rst [hunk 6]
```diff
@@ -206,6 +211,12 @@ Spilling
...
+       spilling io efficiency. The supported compression kind includes 'zlib', 'snappy', 'lzo', 'zstd', 'lz4' and
```
Comment (mbasmanova) [inline velox/docs/configs.rst:218]: "The supported compression kind includes" -> "The supported compression codecs are: ZLIB, SNAPPY,..."

File: velox/docs/configs.rst [hunk 7]
```diff
@@ -206,6 +211,12 @@ Spilling
...
+       'gzip'. If it set to 'none', then there is no compression.g
```
Comment (mbasmanova) [inline velox/docs/configs.rst:219]: Extra 'g' at the end.

"If it set to 'none', then there is no compression" -> "NONE means no compression."

File: velox/exec/tests/utils/OperatorTestBase.cpp [hunk 8]
```diff
@@ -32,6 +32,8 @@
 
 using namespace facebook::velox::common::testutil;
 
+DECLARE_bool(velox_enable_memory_usage_track_in_default_memory_pool);
```
Comment (mbasmanova) [inline velox/exec/tests/utils/OperatorTestBase.cpp:35]: Unrelated change? In general, let's avoid using gflags.

**Review by mbasmanova — COMMENTED** [review]
@xiaoxmeng Thank you for the optimization. Overall looks good. Some minor comments.

File: velox/core/QueryConfig.h (hunk 1 above)
Comment (xiaoxmeng) [reply velox/core/QueryConfig.h:444]: Yeah, it should be 1MB by default. Will tune this later after fix the memory efficiency issue in StreamArena

File: velox/docs/configs.rst (hunk 7 above)
Comment (xiaoxmeng) [reply velox/docs/configs.rst:219]: Thanks!

File: velox/exec/tests/utils/OperatorTestBase.cpp (hunk 8 above)
Comment (xiaoxmeng) [reply velox/exec/tests/utils/OperatorTestBase.cpp:35]: I want to track the spilling memory usage and verify in test. Next will report to Prestissimo for monitoring such as the peak spilling memory usage which might give us some insights on spilling workload and the corresponding memory capacity provision.  I will change this flag to a system config in followup. Thanks!

**Comment by xiaoxmeng** [comment]
> @xiaoxmeng Thank you for the optimization. Overall looks good. Some minor comments.

@mbasmanova thanks for prompt review!

**Comment by mbasmanova** [comment]
@xiaoxmeng There are CI failures. Would you take a look?

File: velox/docs/configs.rst [hunk 9]
```diff
@@ -198,6 +198,11 @@ Spilling
      - integer
      - 0
      - The maximum allowed spill file size. Zero means unlimited.
+   * - spill_write_buffer_size
+     - integer
+     - 4MB
```
Comment (mbasmanova) [inline velox/docs/configs.rst:203]: Should this be 1MB?

File: velox/docs/configs.rst [hunk 10]
```diff
@@ -206,6 +211,12 @@ Spilling
        If the limit is zero, then the spiller always spills a previously spilled partition if it has any data. This is
        to avoid spill from a partition with a small amount of data which might result in generating too many small
        spilled files.
+   * - spill_compression_codec
+     - string
+     - none
+     - Specifies the compression algorithm type to compress the spilled data before write to disk to trade CPU for IO
+       efficiency. The supported compression codecs are: ZLIB, SNAPPY, LZO, ZSTD, LZ4 and GZIP.
```
Comment (mbasmanova) [inline velox/docs/configs.rst:218]: Is this value case sensitive or not? Can I specify NONE or do I have to specify none? Will nOnE work?

File: velox/exec/tests/utils/OperatorTestBase.cpp (hunk 8 above)
Comment (mbasmanova) [reply velox/exec/tests/utils/OperatorTestBase.cpp:35]: Still seeing this gflag. It would be best to avoid introducing new gflags.

**Review by mbasmanova — APPROVED** [review]
Looks good % please remove the gflag.

File: velox/docs/configs.rst (hunk 10 above)
Comment (xiaoxmeng) [reply velox/docs/configs.rst:218]: It is case sensitive but we could make it non-case sensitive in followup.
